### PR TITLE
Caffe Style ImageData 

### PR DESCRIPTION
**Pull request by thuyen**
This `dataset` take a file containing list of images as input. This is very flexible for new datasets since we don't have reorganize images to fit the standard `folder` structure. 

The file content should be similar to `caffe ImageData`:
````
path_to_image1 label1
path_to_image2 label2
path_to_image3 label3
````

### CODE REVIEW

File: torchvision/datasets/image.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import torch.utils.data as data
+from PIL import Image
+import os.path
+
+def make_dataset(file_name):
+    images = []
+    with open(file_name) as f:
```
Comment (alykhantejani) [inline torchvision/datasets/image.py]: The file can be opened in read mode by calling `open(file_name, 'r')`, then you can remove the `read()` below

File: torchvision/datasets/image.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+class ImageData(data.Dataset):
+    """
+    Caffe style ImageData
+
+    Arguments:
+        root: data directory
+        list_file: a file name. Each row of the file: image_path class
+    """
+    def __init__(self, root, list_file, transform=None, target_transform=None,
```
Comment (alykhantejani) [inline torchvision/datasets/image.py]: it would be nice to allow for a default root of `""`, in case your input file contains absolute paths

**Comment by fmassa** [comment]
Hum, looks like this PR provides very similar functionality compared to the [`ListDataset` from `tnt`](https://github.com/pytorch/tnt/blob/master/torchnet/dataset/listdataset.py). Furthermore, I believe that this dataset is not restricted to images, right?

I think it might be a good idea to merge some generic datasets (for example from `tnt`) into core. @apaszke @soumith wdyt?

**Comment by alykhantejani** [comment]
It would also be great if this could support rows of n-elements in the input file (i.e. multiple inputs, multiple outputs) and where outputs are not assumed to be `int`s. For example, we may want floats for regression, or images for examples such as super resolution.

**Comment by fmassa** [comment]
@alykhantejani yes, I agree. ListDataset already supports that, as we can pass a customizable `load` function that receives the content from one line of the list (which could contain an arbitrary number of elements per row)

**Comment by soumith** [comment]
we will be taking some decisions on torchvision next week, including merging open PRs, etc.
please stay tuned and sorry for the delay.

**Comment by yanghuanflc** [comment]
I have already wrote a class called ImageReference which accept a image list file just like caffe ImageData does. If needed I can setup a PR with it.

**Pull request closed without merging**